aesd-char-driver/main.c

```diff
@@ -0,0 +1,303 @@
+#include "aesdchar.h"
+#include <linux/slab.h>	// kmalloc, krealloc, kfree
+
+#ifdef __KERNEL__
+#include <linux/string.h>
+#else
+#include <string.h>
+#endif
+
+
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Ritika Ramchandani"); 
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    PDEBUG("open");
+    
+    struct aesd_dev* dev;
+
+    dev = container_of(inode -> i_cdev, struct aesd_dev, cdev);
+
+    filp -> private_data = dev;
+
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev* dev;
+    size_t entry_byte_offset;
+    struct aesd_buffer_entry *read_entry;
+    size_t num_bytes_read;
+
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+
+    // Check if any of inputs are invalid 
+    if(filp == NULL || buf == NULL)
+    {
+        return -EINVAL;
+    }
+
+    // Get device pointer
+    dev = filp -> private_data;
+
+    // Get interruptible lock
+    if (mutex_lock_interruptible(&(dev -> dev_lock)))
+    {
+        return -ERESTARTSYS;
+    }
+
+    // Get entry at the specified offset from the circular buffer
+    read_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&(dev -> aesd_cb), *f_pos, &entry_byte_offset);
+
+    if(read_entry == NULL)
+    {
+        goto exit_gracefully;
+    }
+
+    // Check if required count is within the bounds of the entry or not
+    if((read_entry -> size - entry_byte_offset) < count)
+    {
+        
+        num_bytes_read = read_entry -> size - entry_byte_offset;
+        
+    }
+    else
+    {
+        num_bytes_read = count;
+    }
+
+    // Copy the entry from the specified offset to the user-provided buffer
+    if(copy_to_user(buf, read_entry -> buffptr + entry_byte_offset, num_bytes_read))
+    {
+        retval = -EFAULT;
+        goto exit_gracefully;
+    }
+
+    *f_pos += num_bytes_read;
+
+    retval = num_bytes_read;
+
+exit_gracefully:
+    mutex_unlock(&dev -> dev_lock);
+    
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    PDEBUG("Write begin\n");
+    
+    ssize_t retval = -ENOMEM;
+    struct aesd_dev* dev;
+    char* temp_buf; 
+    const char* entry_to_rm;
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+
+    // Check if any of inputs are invalid 
+    if(filp == NULL || buf == NULL)
+    {
+        return -EINVAL;
+    }
+
+    // Get device pointer
+    dev = filp -> private_data;
+
+    // Get interruptible lock
+    if (mutex_lock_interruptible(&(dev -> dev_lock)))
+    {
+        return -ERESTARTSYS;
+    }
+		
+    //kmalloc a buffer to copy from user space
+    temp_buf = kmalloc(count, GFP_KERNEL);
```

Comment:
@ritikar97 does this block and the one below it need to be in the critical section?  What is dev->dev_lock locking?